kvserver: move proposal encoding to flush time

tbg opened the pull request:
Previously, the bulk of proposal encoding happened in
`(*Replica).propose` and a buffer representing a part of the encoding
(depending on whether this was a conf change or regular command) was
constructed. This partially encoded proposal would then be handed to the
proposal buffer, which, at flush time, would do "more encoding": it
would complete the marshaled RaftCommand bytes with a "footer"
containing the `MaxLeaseAppliedIndex` and the `ClosedTimestamp`, both of
which are known only at flush time, and would wrap the result in a conf
change proposal, if needed.

This split was awkward. The footer hackery was awkward. It was difficult
to package entry encoding up in a more appropriate subpackage such as
`logstorage`.

This commit moves the entirety of the encoding code away from
`(*Replica).propose` and instead does it when the proposal buffer
flushes. This lets us remove all of the "footer" hackery (since now
we know everything we need to marshal) and collects all of the encoding
in one place, where it will be easier to extract in a future PR.

Epic: CRDB-220
Release note: None

tbg committed:
kvserver: move proposal encoding to flush time

Previously, the bulk of proposal encoding happened in
`(*Replica).propose` and a buffer representing a part of the encoding
(depending on whether this was a conf change or regular command) was
constructed. This partially encoded proposal would then be handed to the
proposal buffer, which, at flush time, would do "more encoding": it
would complete the marshaled RaftCommand bytes with a "footer"
containing the `MaxLeaseAppliedIndex` and the `ClosedTimestamp`, both of
which are known only at flush time, and would wrap the result in a conf
change proposal, if needed.

This split was awkward. The footer hackery was awkward. It was difficult
to package entry encoding up in a more appropriate subpackage such as
`logstorage`.

This commit moves the entirety of the encoding code away from
`(*Replica).propose` and instead does it when the proposal buffer
flushes. This lets us remove all of the "footer" hackery (since now
we know everything we need to marshal) and collects all of the encoding
in one place, where it will be easier to extract in a future PR.

Epic: CRDB-220
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94105)
<!-- Reviewable:end -->

nvb reviewed (commented):
The changes in the last commit make sense to me, especially if they pave the way to encapsulate this all in a library.

My only concern is that this could come with a slight regression for single-range write throughput, as we're losing some parallelism in entry encoding by moving it from each request's evaluation goroutine to the single per-replica raft goroutine. That will slow down the raft goroutine marginally. It also shifts a chunk of memory allocations to the raft goroutine — hopefully GC assist doesn't notice and pick a fight.

Was there anything else in particular that you wanted me to take a look at?

> Reviewed 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94105#-:-NJqkzNA-y_cMKWKamBK:b-c8497t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/replica_proposal_buf.go` line 530 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/94105#-NJqkSrF14lBCWoh0PVv:-NJqkSrF14lBCWoh0PVw:b-dyg6r7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b0d47a29a5adbe04df4752879d9be6205406b58/pkg/kv/kvserver/replica_proposal_buf.go#L530)):*
> ```Go
> 
> 		// TODO(tbg): the encoding code within both of the branches of this
> 		// conditional should be encapsulated in methods in the `logstore` package.
> ```

+1, it's unfortunate how much logic this is adding to the `propBuf`, which is already too complex. But that's not the long-term plan, so it's find for now.
___
*[`pkg/kv/kvserver/kvserverpb/proposer_kv.proto` line 331 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/94105#-NJqjUhp8rhYjcVK3hp1:-NJqjUhp8rhYjcVK3hp2:b-iqjgks) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b0d47a29a5adbe04df4752879d9be6205406b58/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L331)):*
> ```protobuf
>   // marshaled representation with a footer containing the closed timestamp. Due
>   // to a gogoproto quirk[^1] a zero hlc.Timestamp{} still gets put on the wire,
>   // which we wanted to avoid sincei t caused subtle bugs[^2]. None of this
> ```

"since it"


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> My only concern is that this could come with a slight regression for single-range write throughput

That's a good point. I'm tempted to take my chances with this since I find the updated code much clearer, but it's something we should be prepared to address if needed. I think that even if we eventually "walk back" some of this change in the future (i.e. find that we need to shift encoding back to the client goroutines for optimal performance), we'd end up in a better place since we would do so without regressing on the encapsulation of the logic. Speculation, but I'm optimistic.
One "straightforward" thing we could at least do is to pre-allocate a slice using the client goroutine, to shift the allocations off the raft goroutine. I will look into that.

One thing I've wondered about is, now that we are doing all of the marshaling for the batch in one go, could we benefit from actually sharing the backing memory across all slices? That is, if we have ten commands waiting, instead of marshaling then right-sized buffers, could (should) we marshal one right-sized buffer and parcel it out among the individual marshal operations? Presumably, the lifecycle of a collection of entries proposed together is very similar (raft entry cache, etc). This wouldn't really move the needle on the GC assist problem much, so it's more a thought experiment than a practical suggestion.

More somewhat unrelated musings, it would be nice to be able to pool some of the backing memory here. This is difficult right now - `ent.Data` goes into raft, then gets emitted back out as messages, which sit in the raft transport, etc, so this is clearly out of reach now.

> Was there anything else in particular that you wanted me to take a look at?

No, that was very helpful, thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94105#-:-NJtCzIf34YRV8KY69qV:bbof6ii)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: shift entry encoding slice allocation back to proposer

It's better to keep allocations out of the proposal buffer, so that they
don't weigh down the raft processing goroutines.

Epic: CRDB-220
Release note: None

tbg committed:
kvserver: assert that preallocated encoding buffers are sized correctly

I also verified this via pprof but better to check it directly too.

Epic: None
Release note: None

pav-kv reviewed (approved):
I went through it yesterday, dropping some comments/thoughts for now.

It was quite difficult to review `replica_proposal_buf.go` change, so I'll take another look. Right now both before/after are pretty big inlined blocks of code with optimizations (but after is still less optimal because now it's on a critical path?), and the simplicity benefit is largely in TODOs.

I'm not super convinced that it would be too difficult to factor out helpers for 2 phases of encoding the message. Some kind of "prepare"+"finalize" pairs of helpers, maybe done in a type-safe manner so that it's hard to accidentally write/skip a Footer depending on the command type.

I think the following sequence of changes could help the reviewer:
- Factor out encoding helpers (i.e. do the TODOs upfront)
- At this point maybe it's good enough; or the motivation of moving to Flush is more apparent
- If not great, merge "prepare"+"finalize" helpers into one thing, and move invocations from Prepare to the Flush side (but here the diff would be clearer?)

I'll make another pass soon (holidays&stuff), but see if the above makes sense in the meantime.

pavelkalinnikov commented:
Hit Approve accidentally, use with caution :)

pavelkalinnikov commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -456,61 +461,170 @@ func (b *propBuf) FlushLockedWithRaftGroup(
+			ents = append(ents, raftpb.Entry{
+				Data: data,
+			})
+		} else {
```
With the `if-else` construction it is assumed that if the encoding isn't one of the 2 in `if` branch, then it's a config change. This is error-prone if any new encoding is introduced. How about spelling out the encoding explicitly? Maybe use `switch-case` instead. This would also make the first line of the comment below unnecessary.

pavelkalinnikov commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -521,24 +635,14 @@ func (b *propBuf) FlushLockedWithRaftGroup(
 				continue
 			}
-		} else {
-			// Add to the batch of entries that will soon be proposed. It is
```
reflection: Swapping the order of `if` and `else` clauses here made the diff unnecessarily bigger.

pavelkalinnikov commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -456,61 +461,170 @@ func (b *propBuf) FlushLockedWithRaftGroup(
+		case p.command.ReplicatedEvalResult.AddSSTable != nil:
+			entryEncoding = raftlog.EntryEncodingSideloaded
+		case p.command.ReplicatedEvalResult.ChangeReplicas != nil:
+			// Flush any previously batched (non-conf change) proposals to
+			// preserve the correct ordering or proposals. This isn't
+			// strictly necessary for anything - after all, latching prevents
+			// reordering where it matters, but it's not worth changing now.
+			// Later proposals will start a new batch.
+			if err := proposeBatch(raftGroup, b.p.getReplicaID(), ents); err != nil {
+				firstErr = err
+				continue
+			}
```
reflection: Another reason to decouple this from the encoding `switch-case` is: this `proposeBatch()` then wouldn't need to move to a different place, the diff would be cleaner and easier to review.

pavelkalinnikov commented on pkg/kv/kvserver/replica_proposal_buf.go:
```diff
@@ -456,61 +461,170 @@ func (b *propBuf) FlushLockedWithRaftGroup(
+			testingSubmitProposalFilter = filter
+		}
+
+		// TODO(tbg): the encoding code within both of the branches of this
```
Probably a good idea to do it in this PR (not necessarily in `logstore`, it's fine to have a helper in this file, and a TODO to refine and move it elsewhere).

reflection: Not inlining this bit would simplify reviewing the diff in `Flush`.

pav-kv reviewed (approved):
Made another pass over `replica_proposal_buf.go`, it broadly makes sense to me. Left a few "reflection:" comments on how this review could be made easier.

tbg committed:
fixup! kvserver: move proposal encoding to flush time

tbg committed:
kvserver: populate p.command.ClosedTimestamp

This is a nice simplification: we now treat ClosedTimestamp exactly like
MaxLeaseIndex in that it is reflected on the `*RaftCommand`. The fact
that it wasn't bothered me because the `*RaftCommand` is supposed to
reflect the log entry, and here it didn't, breaking an intuitive
invariant. Now it does, and we also get to remove a `sync.Pool`
which is always nice.

Epic: CRDB-220
Release note: None

tbg committed:
[wip] little brain, big dump

tbg committed:
wip
